graylog2-web-interface/src/components/common/SourceCodeEditor.jsx

```diff
@@ -0,0 +1,205 @@
+    this.setState({ selectedText: selectedText });
+  }
+
+  focusEditor = () => {
+    this.reactAce.editor.focus();
+  }
+
+  render() {
+    const { height, width } = this.state;
+    const validCssWidth = Number.isNaN(width) ? '100%' : width;
+    const { theme, resizable } = this.props;
+    const containerStyle = `${style.sourceCodeEditor} ${theme !== 'light' && style.darkMode} ${!resizable && style.static}`;
+    const overlay = <Tooltip id={'paste-button-tooltip'}>Press Ctrl+V (&#8984;V in macOS) or select Edit&thinsp;&rarr;&thinsp;Paste to paste from clipboard.</Tooltip>;
+    return (
+      <div>
+        {this.props.toolbar &&
+          <div className={style.toolbar} style={{ width: validCssWidth }}>
+            <ButtonToolbar>
+              <ButtonGroup>
+                <ClipboardButton title={<i className="fa fa-copy fa-fw" />}
+                                 bsStyle="link"
+                                 bsSize="sm"
+                                 onSuccess={this.focusEditor}
+                                 text={this.state.selectedText}
+                                 buttonTitle="Copy (Ctrl+C / &#8984;C)"
+                                 disabled={this.state.selectedText === ''} />
+                <OverlayTrigger placement="top" trigger="click" overlay={overlay} rootClose>
+                  <Button bsStyle="link" bsSize="sm" title="Paste (Ctrl+V / &#8984;V)">
+                    <i className="fa fa-paste fa-fw" />
+                  </Button>
+                </OverlayTrigger>
+              </ButtonGroup>
+              <ButtonGroup>
+                <Button bsStyle="link"
+                        bsSize="sm"
+                        onClick={this.handleUndo}
+                        title="Undo (Ctrl+Z / &#8984;Z)"
+                        disabled={this.isUndoDisabled()}>
+                  <i className="fa fa-undo fa-fw" />
+                </Button>
+                <Button bsStyle="link"
+                        bsSize="sm"
+                        onClick={this.handleRedo}
+                        title="Redo (Ctrl+Shift+Z / &#8984;&#8679;Z)"
+                        disabled={this.isRedoDisabled()}>
+                  <i className="fa fa-repeat fa-fw" />
+                </Button>
+              </ButtonGroup>
+            </ButtonToolbar>
+          </div>
+        }
+        <Resizable height={height}
+                   width={width}
+                   minConstraints={[200, 200]}
+                   onResize={this.handleResize}>
+          <div className={containerStyle} style={{ height: height, width: validCssWidth }}>
+            <AceEditor ref={(c) => { this.reactAce = c; }}
+                       annotations={this.props.annotations}
+                       editorProps={{ $blockScrolling: 'Infinity' }}
+                       focus={this.props.focus}
+                       fontSize={this.props.fontSize}
+                       mode="text"
```

Comment:
Do we want to extract this into an optional prop, too?
Might come in handy if we ever use this to display/edit JSON or other modes.

Reply:
The main problem is that each mode requires an `import` from `brace`. I couldn't find a way of importing all the modes brace provides, and I'm unsure if it's even safe to do so. I can play a bit more with it and see how it goes. Otherwise I would vote for extending the component when we need it.

Reply:
Yeah makes sense, let's not overcomplicated it for now.